### ISSUE TITLE
[no-floating-promises] flag `.map(async)` result

### ISSUE THREAD

**Issue by rvsit** [typescript-eslint/typescript-eslint#5958]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.8.4&sourceType=module&code=MYewdgzgLgBAJgQygmBeGBtA5FAptLAXQG4AoUxZAOgFsEAHACgQgE8xgZGBKNAPhgBvUjBgIA7ggCWsMLnEwACgCcQNKRFyNGAfWX4ANDH0ArXqgGaoAFSk1cIAK5RGpowEYADN+7cyAXz9SIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6aRAR1ko9oEMA7v0r50URNGgB7aJAA04bHiKlkFanQbM2TabQBm8af3zMA5rWIyAtpWQpxkSTLlKAviHdA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
const data = ['test'];

data.map(async () => {
  await new Promise((_res, rej) => setTimeout(rej, 1000));
});
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/require-await": "error",
    "@typescript-eslint/no-floating-promises": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

Promises must be awaited, end with a call to .catch, end with a call to .then with a rejection handler or be explicitly marked as ignored with the `void` operator.

### Actual Result

No eslint error, but ERR_UNHANDLED_REJECTION when running

### Versions

| package | version |
| -- | -- |
| `@typescript-eslint/eslint-plugin` | `5.42.1` |
| `@typescript-eslint/parser` | `5.42.1` |
| `TypeScript` | `4.8.4` |
| `ESLint` | `8.15.0` |
| `node` | `web` |

**Comment by Josh-Cena** [comment]
Technically speaking, this is not a floating promise, because the return type of `map` is `Array<Promise>`, not `Promise<Array>`. This is why the rule is not able to detect that there's a promise at the top level.

This is the same if you wrap the promise in _anything_:

```js
function getObj(p: Promise<any>) {
  return { p };
}

getObj(Promise.resolve(1));
```

`no-misused-promises` can't catch this either, because that's definitely a legitimate thing to write (passing a promise to another function; returning a promise-containing value from a function).

Let's see if others have ideas; I don't think there's a great heuristic to catch these cases unless we special-case object types and array types.

**Comment by bradzacher** [comment]
I think this would be a pretty good addition to the lint rule!
We could error on this case and suggest wrapping the expression in `await Promise.all(expression)` (could also add suggestions for `.race` and `.allSettled`).

We'd have to put this behind an option (default false for now) to avoid a breaking change, but i'd hazard a guess that we could turn it on by default eventually. Would be good to audit some codebases to be sure.

WDYT @Josh-Cena?

**Comment by Josh-Cena** [comment]
Does that mean we special-case array types and object types? I was talking with @JoshuaKGoldberg and he seemed to dislike special-casing.

**Comment by bradzacher** [comment]
Object types - no I don't think so because there is no construct to handle them natively. So our rule would report and then someone would have to refactor their code completely to extract and await the promise.

But array types are different to object types - they have the native promise methods to await them. This means there's no difficult code change to make - as I mentioned the simplest would be `expression` to `await Promise.all(expression)` - which isn't too different to the non-array case (`await (expression)`)

To me at least handling the array case isn't a special case - instead it's just a natural extension to the rule.

**Comment by JoshuaKGoldberg** [comment]
I'm on the fence here and fine either way. It feels like we're entering _"there's no perfect answer"_ territory. The case of array types being special & built-in makes sense to me but I wouldn't mind either way. 🤷

**Comment by ST-DDT** [comment]
Is this the issue causing the following code to be marked incorrectly as okay?

Fail: `array.map(async () => await foo()).join(',');` // [object Promise],[object Promise]
Correct: `await Promise.all(array.map(() => foo())).join(',')` // value1,value2
